src/Extensions.Hosting.AsyncInitialization/Hosting/AsyncInitializationHostExtensions.cs

```diff
@@ -18,19 +18,78 @@ public static class AsyncInitializationHostExtensions
+        /// <param name="cancellationToken">Optionally propagates notifications that the operation should be cancelled</param>
+        /// <returns>The <see cref="Task"/> that represents the asynchronous operation.</returns>
+        /// <exception cref="ArgumentNullException">Thrown when the host is null.</exception>
+        /// <exception cref="InvalidOperationException">Thrown when the initialization service has not been registered.</exception>
+        /// <exception cref="OperationCanceledException">Thrown when the cancellationToken is cancelled.</exception>
+        public static async Task InitAndRunAsync(this IHost host, CancellationToken cancellationToken = default)
+        {
+            if (host == null) throw new ArgumentNullException(nameof(host));
+
+            await using var scope = host.Services.CreateAsyncScope();
+            var rootInitializer = scope.ServiceProvider.GetService<RootInitializer>()
+                ?? throw new InvalidOperationException("The async initialization service isn't registered, register it by calling AddAsyncInitialization() on the service collection or by adding an async initializer.");
+
+
+            await using (host.AsAsyncDisposable().ConfigureAwait(false))
             {
-                throw new InvalidOperationException("The async initialization service isn't registered, register it by calling AddAsyncInitialization() on the service collection or by adding an async initializer.");
+                try
+                {
+                    await rootInitializer.InitializeAsync(cancellationToken).ConfigureAwait(false);
+                    await host.StartAsync(cancellationToken).ConfigureAwait(false);
+                    await host.WaitForShutdownAsync(cancellationToken).ConfigureAwait(false);
+                }
+                finally
+                {
+                    await rootInitializer.TeardownAsync(CancellationToken.None).ConfigureAwait(false);
+                }
             }
+        }
 
-            await rootInitializer.InitializeAsync(cancellationToken);
+        // wraps an IHost instance in an IAsyncDisposable 
+        private static IAsyncDisposable AsAsyncDisposable(this IHost host)
+        {
+            return host as IAsyncDisposable ?? new AsyncDisposableHostWrapper(host);
+        }
+
+        // IAsyncDisposable wrapper for IHost
+        private class AsyncDisposableHostWrapper : IAsyncDisposable
+        {
+            private readonly IHost _host;
+            public AsyncDisposableHostWrapper(IHost host)
+            {
+                _host = host ?? throw new ArgumentNullException(nameof(host));
+            }
+            public ValueTask DisposeAsync()
+            {
+                if (_host is IAsyncDisposable asyncDisposable)
+                {
+                    return asyncDisposable.DisposeAsync();
+                }
+                _host.Dispose();
+                return default;
```

Comment:
Is `default` the same as `ValueTask.CompletedTask`?
It's probably fine, but it surprised me when I saw it... Using `ValueTask.CompletedTask` would probably be clearer.